orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/graph/v2/transform/V2ConfigStageInjectionTransform.java

```diff
@@ -0,0 +1,300 @@
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.pipelinetemplate.v1schema.graph.v2.transform;
+
+import com.netflix.spinnaker.orca.pipelinetemplate.exceptions.IllegalTemplateConfigurationException;
+import com.netflix.spinnaker.orca.pipelinetemplate.v1schema.model.StageDefinition;
+import com.netflix.spinnaker.orca.pipelinetemplate.v2schema.V2PipelineTemplateVisitor;
+import com.netflix.spinnaker.orca.pipelinetemplate.v2schema.model.V2PipelineTemplate;
+import com.netflix.spinnaker.orca.pipelinetemplate.v2schema.model.V2StageDefinition;
+import com.netflix.spinnaker.orca.pipelinetemplate.v2schema.model.V2TemplateConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+
+public class V2ConfigStageInjectionTransform implements V2PipelineTemplateVisitor {
+
+  private final Logger log = LoggerFactory.getLogger(getClass());
+  private static final Logger slog = LoggerFactory.getLogger(V2ConfigStageInjectionTransform.class);
+
+  private V2TemplateConfiguration templateConfiguration;
+
+  public V2ConfigStageInjectionTransform(V2TemplateConfiguration templateConfiguration) {
+    this.templateConfiguration = templateConfiguration;
+  }
+
+  @Override
+  public void visitPipelineTemplate(V2PipelineTemplate pipelineTemplate) {
+    replaceStages(pipelineTemplate);
+    injectStages(pipelineTemplate);
+    log.info(",, post stage injection template: {}", pipelineTemplate);
+  }
+
+  private void replaceStages(V2PipelineTemplate pipelineTemplate) {
+    List<V2StageDefinition> templateStages = pipelineTemplate.getStages();
+    for (V2StageDefinition confStage : templateConfiguration.getStages()) {
+      for (int i = 0; i < templateStages.size(); i++) {
+        if (templateStages.get(i).getRefId().equals(confStage.getRefId())) {
+          templateStages.set(i, confStage);
+          log.debug("Template '{}' stage '{}' replaced by config {}", pipelineTemplate.getId(), confStage.getRefId(), templateConfiguration.getRuntimeId());
+        }
+      }
+    }
+  }
+
+  private void injectStages(V2PipelineTemplate pipelineTemplate) {
+    // Create initial graph via dependsOn. We need to include stages that the configuration defines with dependsOn as well
+    List<V2StageDefinition> initialStages = pipelineTemplate.getStages();
+
+    initialStages.addAll(templateConfiguration.getStages().stream()
+      .filter(s -> !s.getRequisiteStageRefIds().isEmpty())
+      .collect(Collectors.toList()));
+
+    pipelineTemplate.setStages(
+      createGraph(initialStages)
+    );
+
+    // Handle stage injections.
+    List<V2StageDefinition> templateInjectStages = injectStages(pipelineTemplate.getStages(), pipelineTemplate.getStages());
+    List<V2StageDefinition> configInjectStages = injectStages(templateConfiguration.getStages(), templateInjectStages);
+    pipelineTemplate.setStages(configInjectStages);
+  }
+
+  private enum Status {
```

Comment:
Do you think there will be more states? If not, maybe a boolean value is more appropriate?

Reply:
This logic was hardened in MPT v1 aside from the `V2StageDefinition` changes, so I'm inclined to leave this implementation as is.

Reply:
Still adding tests though.